Fix nested volume mounts for read-only API data volumes

joelsmith opened the pull request:
**What this PR does / why we need it**:

Since the runtime may try to create mount points within the sandbox, it will fail if the mount point is within a read-only API data volume, like a secret or configMap volume.

Create any needed mount points during volume setup.

**Which issue(s) this PR fixes**:
Fixes #60341

**Special notes for your reviewer**:


**Release note**:

No release note needed since we haven't released any versions with this bug present (yet)

```release-note
NONE

```

joelsmith committed:
Fix nested volume mounts for read-only API data volumes

Since the runtime may try to create mount points within
the sandbox, it will fail if the mount point is within
a read-only API data volume, like a secret or configMap
volume.

Create any needed mount points during volume setup.

joelsmith commented:
/sig storage
/sig node
/kind bug

joelsmith commented:
@sjenning @jsafrane PTAL

joelsmith commented:
Weird, pull-kubernetes-e2e-gce shows up as still running on its status page, but it shows up as failed here. Something must have gone wrong with it. https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/60342/pull-kubernetes-e2e-gce/22842/
/test pull-kubernetes-e2e-gce

jsafrane commented on pkg/volume/util/nested_volumes_test.go:
```diff
@@ -0,0 +1,103 @@
+/*
+Copyright 2016 The Kubernetes Authors.
```
2018

jsafrane commented on pkg/volume/util/nested_volumes.go:
```diff
@@ -0,0 +1,66 @@
+		var mountPoints []string
+		var myMountPoints []string
+		for _, vol := range container.VolumeMounts {
+			mountPoints = append(mountPoints, filepath.Clean(vol.MountPath))
```
Use a variable for `filepath.Clean(vol.MountPath)` so you don't need to do `mountPoints[len(mountPoints)-1]` two lines below.

jsafrane commented on pkg/volume/util/nested_volumes.go:
```diff
@@ -0,0 +1,66 @@
+		parentPrefix := ".." + string(os.PathSeparator)
+		for _, mountPoint := range myMountPoints {
+			if strings.HasPrefix(mountPoint, parentPrefix) {
+				continue // Don't let a container trick us into creating directories outside of its rootfs
```
Should this return an error?

joelsmith commented:
Thanks @jsafrane . I have updated with the items from your review and I'll squash once everybody's happy.

jsafrane commented:
/approve
I'd like second par of eyes to look at it. This is tricky code. In general, volume plugins should not care about other volumes...

joelsmith commented:
Yeah, it's unfortunate that these volumes have to care about other volumes. Other read-only mounted volumes like PVs can have their nested mount points created via other means so that Docker doesn't try (and fail) to create missing mount points.

sjenning commented on pkg/volume/util/nested_volumes.go:
```diff
@@ -0,0 +1,68 @@
+
+// MakeNestedMountpoints creates mount points in baseDir for volumes mounted beneath name
+func MakeNestedMountpoints(name, baseDir string, pod v1.Pod) error {
+	for i := 0; i < len(pod.Spec.InitContainers)+len(pod.Spec.Containers); i++ {
```
Took me a minute to figure out what this did.  Can we break line 38+ in to separate function and just have to "for-range loops here to iterate over the Containers and InitContainers.  The index math is giving me flashbacks of my kernel dev days!

sjenning commented on pkg/volume/util/nested_volumes.go:
```diff
@@ -0,0 +1,68 @@
+				return fmt.Errorf("Invalid container mount point %v", mountPoint)
+			}
+			mpSlash := mountPoint + string(os.PathSeparator)
+			for prev, cur := -1, sort.SearchStrings(mountPoints, mpSlash); cur < len(mountPoints) && strings.HasPrefix(mountPoints[cur], mpSlash); cur++ {
```
This code block needs some big time comments. It is super icky :-/

joelsmith replied:
I've added comments, but maybe it would be better also to refactor it. Do you have any ideas on a cleaner impl of equivalent logic?

sjenning replied:
Let me try to make time to understand exactly what is happening here now that you've added the comments.  Try to reply today.

msau42 commented on pkg/volume/util/nested_volumes.go:
```diff
@@ -0,0 +1,78 @@
+func MakeNestedMountpoints(name, baseDir string, pod v1.Pod) error {
+	checkContainer := func(container *v1.Container) error {
+		var mountPoints []string
+		var myMountPoints []string
```
Can you add a comment that this is the mount point for the volume with 'name'

msau42 replied:
Also can you add a test case where you have multiple mount points for the same volume name?

msau42 commented on pkg/volume/util/nested_volumes.go:
```diff
@@ -0,0 +1,78 @@
+			// The loop continues until it reaches the end of mountPoints or until cur references a mount point that is not nested beneath mountPoint
+			// prev represents the the last-encountered nested mountPoint. If set (i.e., not -1), it's used to compare with a mount point to make
+			//      sure the mount point is not a double-nested mount point
+			for prev, cur := -1, sort.SearchStrings(mountPoints, mpSlash); cur < len(mountPoints) && strings.HasPrefix(mountPoints[cur], mpSlash); cur++ {
```
Would it be better to create a new filtered mount points list where all the entries have mpSlash as the prefix?

Could you potentially have a scenario where:
* mpSlash is in the mount point string, but is not the prefix
* that mount point comes before the real mpSlash mount points in sorted order, so this loop will break out early

joelsmith replied:
I'm not certain I understand your scenario, but if I do, I don't think it is possible only because mpSlash necessarily has a trailing slash.  So, for example, given `/mnt`, `/mnt/nested` and `/mnt_not_nested`, when `/mnt` is being prepared, the `SearchStrings()` call would return the first thing after `/mnt/`, which would be `/mnt/nested`. This is true whether `_` or `/` comes first lexicographically because of the appended `/`.  With that in mind, do you still think it's worth filtering the list before looping? (I'm fine either way).

msau42 replied:
Ah sorry I thought SearchStrings would return the first string where it matched by subset.  But it does exact match, so this is fine.

msau42 commented on pkg/volume/util/nested_volumes.go:
```diff
@@ -0,0 +1,78 @@
+				myMountPoints = append(myMountPoints, cleaned)
+			}
+		}
+		sort.Strings(mountPoints)
```
Test case for when sorting would actually matter

joelsmith replied:
Added

msau42 replied:
There are some characters that come after "/" in sorted order:
```
! " $ % & ' ( ) * , - . / < > ? @ [ ] ^ { } ~
```

I haven't been able to come up with a scenario where it actually matters though.

joelsmith commented:
@sjenning @msau42 I have refactored the comparison loop and renamed a few variables in the hopes that it's a little clearer. Please let me know if you prefer the old or the new version or a variant of one of those (or something else entirely!).

joelsmith commented:
Looks like we hit flakes #59426 and #60447 
/retest

msau42 commented on pkg/volume/util/nested_volumes.go:
```diff
@@ -0,0 +1,96 @@
+	"strings"
+)
+
+func getNestedMountpoints(name, baseDir string, pod v1.Pod) ([]string, error) {
```
Add a comment that this returns nested mountpoints relative to baseDir

msau42 reviewed (commented):
I like this better, much easier to read and understand.

ixdy commented:
/retest

sjenning commented:
Thanks for the changes
/lgtm

joelsmith commented:
Why is `cla/linuxfoundation` stuck? Can I trigger it to retry?
/test cla/linuxfoundation

joelsmith commented:
It got knocked out of the submit queue.
/retest
